framework/sandstone.cpp

```diff
@@ -1042,6 +1047,87 @@ int num_packages() {
 }
 
 namespace {
+class BarrierDeviceSchedule : public DeviceSchedule
+{
+public:
+    void device_reschedule() override {
+        auto on_completion = [&]() noexcept {
+            std::unique_lock lock(groups_mutex);
+
+            int group_idx = thread_num / members_per_group;
+            GroupInfo *group = groups[group_idx];
+
+            // Rotate cpus vector so reschedule group members to a different cpu
+            std::rotate(group->next_cpu.begin(), group->next_cpu.begin() + 1, group->next_cpu.end());
+            lock.unlock();
+
+            // Reschedule group members
+            for (int i=0; i<members_per_group; i++) {
+                if(!pin_thread_to_logical_processor(LogicalProcessor(cpu_info[group->next_cpu[i]].cpu_number), group->pid[i]))
+                    log_debug("Failed to reschedule %d (%lld) to CPU %d", thread_num, pthread_self(), cpu_info[group->next_cpu[i]].cpu_number);
+            }
+        };
+
+        std::unique_lock lock(groups_mutex);
+
+        // Initialize groups on first run
+        if (groups.size() == 0) {
+            groups.reserve(num_cpus()/members_per_group);
+            for (int i=0; i<num_cpus(); i+=members_per_group) {
+                groups.push_back(new GroupInfo(members_per_group, on_completion));
+            }
+        }
+
+        // Fill thread info if not done already
+        int group_idx = thread_num / members_per_group;
+        GroupInfo *group = groups[group_idx];
+        int thread_info_idx = thread_num % members_per_group;
+        if (group->pid[thread_info_idx] == 0) {
+            group->pid[thread_info_idx] = getpid();
```

Comment:
`getpid()` is probably wrong here. Did you mean `gettid()` and thus use a  `tid_t`?

We've discussed (but haven't implemented) the framework knowing and storing the `tid_t` for each thread it starts, so you don't have to do it here. That way, the `pin_thread_to_logical_processor` function would take an internal CPU number instead of a `tid_t`. But let's think about that with a later patch.

Reply:
Storing the TID was implemented some time ago in commit 726f637, but I kind find it useless for this because `sched_setaffinity`, on `pin_thread_to_logical_processor()`, expects a PID not a TID.